File: typescript/Organization.ts
import Position from './Position';
import Name from './Name';
import Employee from './Employee';

abstract class Organization {
  private root: Position;
  EmpId: number;

  constructor() {
    this.root = this.createOrganization();
  }

  protected abstract createOrganization(): Position;

  printOrganization = (position: Position, prefix: string): string => {
    let str = `${prefix}+-${position}\n`;
    for (const p of position.getDirectReports()) {
      str = str.concat(this.printOrganization(p, `${prefix}  `));
    }
    return str;
  };

  // Hire the given person as an employee in the position that has that title
  // Return the newly filled position or undefined if no position has that title
  hire(person: Name, title: string): Position | undefined {
    
    const emp=new Employee(this.EmpId,person);
    const pos=new Position(title,emp);
    console.log(pos.toString());
    this.EmpId++;
    return pos;
 
  }

  toString = () => this.printOrganization(this.root, '');
}

export default Organization;

File: java/com/aa/act/interview/org/Organization.java
package com.aa.act.interview.org;

import java.util.Optional;

public abstract class Organization {

	private Position root;
	private int EmpId = 1;
	
	public Organization() {
		root = createOrganization();
	}
	
	protected abstract Position createOrganization();
	
	/**
	 * hire the given person as an employee in the position that has that title
	 * 
	 * @param person
	 * @param title
	 * @return the newly filled position or empty if no position has that title
	 */
	public Optional<Position> hire(Name person, String title) {
		
				// created an instance of employee object expecting an identifier and a string.
				Employee e=new Employee(EmpId, person);

				//setting position by passing employee object and title of the employee.
        		Position p=new Position(title, e);
                System.out.println("Newly Hired Employee is "+p.toString());
                EmpId++;
		return Optional.of(p);
	}

	@Override
	public String toString() {
		return printOrganization(root, "");
	}
	
	private String printOrganization(Position pos, String prefix) {
		StringBuffer sb = new StringBuffer(prefix + "+-" + pos.toString() + "\n");
		for(Position p : pos.getDirectReports()) {
			sb.append(printOrganization(p, prefix + "\t"));
		}
		return sb.toString();
	}
}
